k2/csrc/cuda/debug_test.cu

```diff
@@ -0,0 +1,190 @@
+// k2/csrc/cuda/debug_test.cu
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#ifndef K2_PARANOID
+#define K2_PARANOID
+#endif
+
+#include "k2/csrc/cuda/debug.h"
+
+namespace k2 {
+
+// A dummy cuda asynchronous function
+__global__ void FillContents(int n, int *output) {
+  int correct_index = static_cast<int>(threadIdx.x) * n;
+
+  for (int i = correct_index; i < n; i++) {
+    output[i] = (i - correct_index) * 2;
+    K2_DLOG("Print from async cuda kernel, output[%d] = %d\n", i, output[i]);
+  }
+}
+
+/* number of blocks (max 65,536) */
+#define BLOCKS 50000
+
+/* number of threads per block (max 1,024) */
+#define THREADS 200
+
+/* N is now the number of blocks times the number of threads
+   in this case it is 10 million */
+#define N (BLOCKS * THREADS)
+
+/* the kernel fills an array up with square roots */
+__global__ void square_roots(double* array) {
+  /* find my array index
+     this now uses three values:
+     blockIdx.x which is the block ID we are on
+     blockDim.x which is the number of blocks in the X dimension
+     threadIdx.x which is the thread ID we are */
+  unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+  /* compute the square root of this number and store it in the array
+     CUDA provides a sqrt function as part of its math library, so this
+     call uses that - not the one from math.h which is a CPU function */
+  array[idx] = sqrt((double) idx);
+}
+
+__global__ void HelloCUDA(float f) {
+  if (threadIdx.x == 0)
+    K2_DLOG("Hello thread %d, f=%f\n", threadIdx.x, f);
+}
+
+// A vector add kernel definition
+__global__ void VecAdd(float* A, float* B, float* C) {
+  int i = threadIdx.x;
+  C[i] = A[i] + B[i];
+  K2_PARANOID_ASSERT(C[i] == A[i] + B[i],
+                     "Error: C[%d] is %f != A[%d] (%f) + B[%d] (%f)\n",
+                     i, C[i], i, B[i], i, B[i]);
+}
+
+TEST(DebugTest, StaticAssert) {
+  K2_STATIC_ASSERT(9 % 3 == 0, "The value is unexpected");
+}
+
+TEST(DebugTest, K2Assert) {
+  K2_ASSERT(2 > 1);
+  ASSERT_DEATH(K2_ASSERT(2 < 1), "");
+}
+
+TEST(DebugTest, K2CheckEq) {
+  K2_CHECK_EQ(2 + 3, 5);
+}
+
+TEST(DebugTest, K2CudaCheckError) {
+  ASSERT_DEATH(K2_CUDA_CHECK_ERROR(cudaErrorMemoryAllocation),
+               "cudaErrorMemoryAllocation");
+}
+
+TEST(DebugTest, K2CudaSafeCall) {
+
+  /* store the square roots of 0 to (N-1) on the CPU
+   * stored on the heap since it's too big for the stack for large values of N */
+  double* roots = (double*) malloc(N * sizeof(double));
```

Comment:
1. Please use `reinterpret_cast`.
2. This is c++ code and there is no need to follow the c style. Once again,
I would suggest you using either `new double[N]` or `std::vector<double> roots(N)`;

Reply:
Hmm, I don't want to keep change this kind of what's your favorite thing. And it's not cpp code as its filename indicate, and the target error comes from device.

Reply:
this is about code style; it can help you reduce the number of programming errors, e.g., memory leaks.

Anyway, you can leave it as is; my comments are just advice.

---

By the way, I find that a large part of your code (even the comment) is just copying and pasting from someplace.
That may explain why your code contains different styles. Please leave a notice indicating where the code is from if it is copied verbatim.